core/src/main/java/io/questdb/griffin/engine/functions/groupby/CountDistinctIntGroupByFunction.java

```diff
@@ -48,41 +56,43 @@ public CountDistinctIntGroupByFunction(Function arg) {
 
     @Override
     public void clear() {
-        sets.clear();
-        setIndex = 0;
+        groupId = 0;
+        groupedValues.restoreInitialCapacity();
     }
 
     @Override
-    public void computeFirst(MapValue mapValue, Record record) {
-        final IntHashSet set;
-        if (sets.size() <= setIndex) {
-            sets.extendAndSet(setIndex, set = new IntHashSet());
-        } else {
-            set = sets.getQuick(setIndex);
-        }
+    public void close() {
+        UnaryFunction.super.close();
+        groupId = 0;
+        Misc.free(groupedValues);
+    }
 
-        set.clear();
+    @Override
+    public void computeFirst(MapValue mapValue, Record record) {
         final int val = arg.getInt(record);
         if (val != Numbers.INT_NaN) {
-            set.add(val);
-            mapValue.putLong(valueIndex, 1L);
+            MapKey groupedMapKey = groupedValues.withKey();
+            groupedMapKey.putInt(groupId);
+            groupedMapKey.putInt(val);
+            groupedMapKey.createValue();
+            mapValue.putLong(valueIndex, 1);
```

Comment:
Can it be `putBoolean()` instead to reduce the memory?

Reply:
This `valueIndex` map column stores the `count_distinct()` result, so the `long` counter. Are you referring to another line or I'm missing something?